File: Backtracking/wordsearch.cpp
#include <bits/stdc++.h>
using namespace std;

bool exist(vector<vector<char>> &board, string word)
{
    int row = board.size();
    int col = board[0].size();

    for (int i = 0; i < row; i++)
    {
        for (int j = 0; j < col; j++)
        {
            if (word[0] == board[i][j])
            {
                if (dfs(board, word, 0, i, j, row, col))
                {
                    return true;
                }
            }
        }
    }
}

bool dfs(vector<vector<char>> &board, string word, int idx, int i, int j, int row, int col)
{
    if (i < 0 || i >= row || j < 0 || j >= col || board[i][j] != word[idx])
    {
        return false;
    }

    if (idx == word.size() - 1)
    {
        return true;
    }

    board[i][j] = '#';

    if (dfs(board, word, idx + 1, i + 1, j, row, col) || dfs(board, word, idx + 1, i - 1, j, row, col) || dfs(board, word, idx + 1, i, j + 1, row, col) || dfs(board, word, idx + 1, i, j - 1, row, col))
    {
        return true;
    }

    board[i][j] = word[idx];

    return false;
}

File: Backtracking/wordSearch.py
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        self.row, self.col = len(board), len(board[0])
        path = set()

        def backtracking(i, j, index):
            if index == len(word):
                return True

            if (
                i < 0
                or j < 0
                or i >= self.row
                or j >= self.col
                or board[i][j] != word[index]
                or (i, j) in path
            ):
                return False

            path.add((i, j))

            if (
                backtracking(i + 1, j, index + 1)
                or backtracking(i - 1, j, index + 1)
                or backtracking(i, j + 1, index + 1)
                or backtracking(i, j - 1, index + 1)
            ):
                return True

            path.remove((i, j))

            return False

        for i in range(self.row):
            for j in range(self.col):
                if board[i][j] == word[0]:
                    if backtracking(i, j, 0) == True:
                        return True

        return False
